### PR TITLE
[fix] #2420: Fix "Permission already exists" bug

### LINKED ISSUE #2420: Fix `Permission already exists` bug
We've encountered a bug: Iroha rejects transactions with permission granting, saying, that it already exists for given user, when it's actually not.

I've found, that the problem is in how `PermissionToken` is sorted in `BTreeSet`. More particularly the problem is in `PartialOrd` and `Ord` implementation for `PermissionToken`.

Need to fix it and write some tests to check if it works

### PR DESCRIPTION
**Pull request by Arjentix**
### Description of the Change

- `PermissonToken` now derives `PartialOrd` and `Ord`. Manual implementation that was using only token name used to cause collision and is the reason of the bug
- `PartialOrd` and `Ord` were derived for many other structs that `PermissionToken` depends on
- Added test to check this bug

### One more *dangerous* bug was found and fixed with the same changes

While fixing this bug I have also found another bug which this PR also fixes.
`PermissionToken` collision used to make it possible to i.e. have a permission to modify only one asset, but the same permission was enough to modify any other asset.
The test I introduced checks that too.

### Issue

- Closes #2420 

### Benefits

- Permission collision bug fixed
- One more test

### Possible Drawbacks

None

### CODE REVIEW

**Review by mversic — COMMENTED** [review]
I recently removed `PartialOrd` and `Ord` from most of these structures. What does it mean that you can order `Equal` or `Subtract` or `SignatureConditionCheck` and alike? Also, once you make `Value` to be `Ord` it just forces every type it wraps to be as well. Does it need that requirement?

I don't think I touched the `Ord` implementation of `PermissionToken` - that one made sense.

File: data_model/src/lib.rs [hunk 1]
```diff
@@ -286,7 +286,18 @@ pub type ValueBox = Box<Value>;
...
+    Clone,
+    PartialEq,
+    Eq,
+    Decode,
+    Encode,
+    Deserialize,
+    Serialize,
+    FromVariant,
+    IntoSchema,
+    PartialOrd,
```
Comment (appetrosyan) [inline data_model/src/lib.rs:299]: I'm not sure if total ordering of the catch-all Value type makes sense, but I'm guessing that we need this for another upstream type?

**Comment by mversic** [comment]
my argument for removing them was that they are somewhat of a cheat/lie in many places even though they can, formally speaking, be ordered. Anyways, I don't mind if they stay

**Comment by Arjentix** [comment]
Yeah, making value to be `Ord` was quite painful. But how can we properly implement `Ord` for `PermissionToken` without it?
I assume, we can use HashSet instead of BTreeSet. Then requirements will be different.
But I don't see a problem with `Ord` for now

File: data_model/src/permissions.rs [hunk 2]
```diff
@@ -34,6 +34,8 @@ pub type Permissions = btree_set::BTreeSet<PermissionToken>;
     Deserialize,
     Serialize,
     IntoSchema,
+    PartialOrd,
```
Comment (Erigara) [inline data_model/src/permissions.rs:37]: Crucial part, cmp not only by name, but also by params.

**Comment by Arjentix** [comment]
Or we can force `PermissionToken` to store `String` or `Name` instead of `Value`

**Review by mversic — APPROVED** [review]

**Review by Erigara — APPROVED** [review]

**Comment by Arjentix** [comment]
> my argument for removing them was that they are somewhat of a cheat/lie in many places even though they can, formally speaking, be ordered

I agree, there is a little smell... But it looks like the simplest solution. We can refactor it after the merge window closes. But I don't know what we can change here to make it working

**Comment by Erigara** [comment]
> > my argument for removing them was that they are somewhat of a cheat/lie in many places even though they can, formally speaking, be ordered
> 
> I agree, there is a little smell... But it looks like the simplest solution. We can refactor it after the merge window closes. But I don't know what we can change here to make it working

same feeling, not really make sense to compare `Equal`s, but most straightforward solution

File: client/tests/integration/permissions.rs [hunk 3]
```diff
@@ -201,3 +209,83 @@ fn permissions_checked_before_transaction_execution() {
 
     assert!(root_cause.contains("Account does not have the needed permission token"));
 }
+
+#[test]
+fn permissions_differ_not_only_by_names() {
```
Comment (appetrosyan) [inline client/tests/integration/permissions.rs:214]: Good testing style: a scenario is more important to test than specific behaviour.

**Review by appetrosyan — APPROVED** [review]

**Pull request merged**